homeassistant/components/aosmith/config_flow.py

```diff
@@ -0,0 +1,110 @@
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import aiohttp_client
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required("email"): str,
+        vol.Required("password"): str,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    client = AOSmithAPIClient(data["email"], data["password"], websession)
+
+    try:
+        await client.get_devices()
+    except AOSmithInvalidCredentialsException:
+        raise InvalidAuth
+
+    return {"title": data["email"]}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for A. O. Smith."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Start the config flow."""
+        self._reauth_entry = None
+        self._email = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            unique_id = user_input["email"].lower()
+            await self.async_set_unique_id(unique_id)
+            if not self._reauth_entry:
+                self._abort_if_unique_id_configured()
+
+            try:
+                info = await validate_input(self.hass, user_input)
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                if not self._reauth_entry:
+                    return self.async_create_entry(title=info["title"], data=user_input)
+
+                self.hass.config_entries.async_update_entry(
+                    self._reauth_entry, data=user_input, unique_id=unique_id
+                )
+
+                # Reload the config entry otherwise devices will remain unavailable
+                self.hass.async_create_task(
+                    self.hass.config_entries.async_reload(self._reauth_entry.entry_id)
+                )
+                return self.async_abort(reason="reauth_successful")
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required("email", default=self._email): str,
+                    vol.Required("password"): str,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
```

Comment:
Please add reauth in a follow up PR

Reply:
Is this a requirement? It doesn't add much complexity and I'd really prefer to have it in this PR if possible. But if this is a blocker to merging then I'm ok with moving it out of this PR. When I [previously submitted an integration](https://github.com/home-assistant/core/pull/45768), having reauth in the initial PR wasn't an issue.

Reply:
I think I am going to put this as a requirement. Google Tasks also didn't come with a reauth flow. I think this PR looks quite done, I'll take a look at the tests now. After this PR you can add the reauth flow and I'll happily take a look at it

Reply:
Removed reauth from this PR. Will submit in a follow-up PR.